File: backend-go/utils/settings_store.go
package utils

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"file-lite-go/config"
)

type SettingsStoreValue = any

type settingsStoreMap map[string]SettingsStoreValue

type settingsStoreState struct {
	mu     sync.Mutex
	loaded bool
	cache  settingsStoreMap
}

var frontendSettingsStore = &settingsStoreState{}

func GetSettingsValue(key string) (SettingsStoreValue, error) {
	frontendSettingsStore.mu.Lock()
	defer frontendSettingsStore.mu.Unlock()

	if err := frontendSettingsStore.ensureLoaded(); err != nil {
		return nil, err
	}
	if value, ok := frontendSettingsStore.cache[key]; ok {
		return value, nil
	}
	return nil, nil
}

func SetSettingsValue(key string, value SettingsStoreValue) (SettingsStoreValue, error) {
	frontendSettingsStore.mu.Lock()
	defer frontendSettingsStore.mu.Unlock()

	if err := frontendSettingsStore.ensureLoaded(); err != nil {
		return nil, err
	}

	nextStore := cloneSettingsStore(frontendSettingsStore.cache)
	nextStore[key] = value
	if err := persistSettingsStore(nextStore); err != nil {
		return nil, err
	}
	frontendSettingsStore.cache = nextStore
	return value, nil
}

func DeleteSettingsValue(key string) (nilValue SettingsStoreValue, err error) {
	frontendSettingsStore.mu.Lock()
	defer frontendSettingsStore.mu.Unlock()

	if err := frontendSettingsStore.ensureLoaded(); err != nil {
		return nil, err
	}
	if _, ok := frontendSettingsStore.cache[key]; !ok {
		return nil, nil
	}

	nextStore := cloneSettingsStore(frontendSettingsStore.cache)
	delete(nextStore, key)
	if err := persistSettingsStore(nextStore); err != nil {
		return nil, err
	}
	frontendSettingsStore.cache = nextStore
	return nil, nil
}

func (s *settingsStoreState) ensureLoaded() error {
	if s.loaded {
		return nil
	}
	store, err := readSettingsStoreFile()
	if err != nil {
		return err
	}
	s.cache = store
	s.loaded = true
	return nil
}

func readSettingsStoreFile() (settingsStoreMap, error) {
	content, err := os.ReadFile(config.FrontendStorageFilePath())
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return settingsStoreMap{}, nil
		}
		return nil, fmt.Errorf("read frontend settings store: %w", err)
	}

	var value any
	if err := json.Unmarshal(content, &value); err != nil {
		return settingsStoreMap{}, nil
	}
	return normalizeSettingsStoreMap(value), nil
}

func normalizeSettingsStoreMap(value any) settingsStoreMap {
	if value == nil {
		return settingsStoreMap{}
	}
	store, ok := value.(map[string]any)
	if !ok {
		return settingsStoreMap{}
	}
	return settingsStoreMap(store)
}

func persistSettingsStore(store settingsStoreMap) error {
	filePath := config.FrontendStorageFilePath()
	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
		return fmt.Errorf("create frontend settings dir: %w", err)
	}
	content, err := json.MarshalIndent(store, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal frontend settings store: %w", err)
	}
	if err := os.WriteFile(filePath, content, 0644); err != nil {
		return fmt.Errorf("write frontend settings store: %w", err)
	}
	return nil
}

func cloneSettingsStore(store settingsStoreMap) settingsStoreMap {
	cloned := make(settingsStoreMap, len(store))
	for key, value := range store {
		cloned[key] = value
	}
	return cloned
}

File: backend/src/utils/settings-store.ts
import fs from 'node:fs/promises'
import Path from 'node:path'
import { getFrontendStorageFilePath } from '@/config/config.ts'

export type SettingsStoreValue
  = | null
    | boolean
    | number
    | string
    | SettingsStoreValue[]
    | { [key: string]: SettingsStoreValue }

type SettingsStoreMap = Record<string, SettingsStoreValue>

export interface ReloadedSettingsStore {
  previous: SettingsStoreMap
  current: SettingsStoreMap
}

let settingsStoreCache: SettingsStoreMap | null = null
let settingsStoreLoadPromise: Promise<SettingsStoreMap> | null = null
let settingsStoreWriteQueue: Promise<void> = Promise.resolve()

function normalizeSettingsStoreMap(value: unknown): SettingsStoreMap {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return {}
  }

  return Object.fromEntries(Object.entries(value)) as SettingsStoreMap
}

async function readSettingsStoreFile(): Promise<SettingsStoreMap> {
  try {
    const content = await fs.readFile(getFrontendStorageFilePath(), 'utf-8')
    return normalizeSettingsStoreMap(JSON.parse(content))
  }
  catch (error: any) {
    if (error?.code === 'ENOENT') {
      return {}
    }
    console.error('Failed to read frontend settings store', error)
    return {}
  }
}

async function ensureSettingsStoreLoaded(): Promise<SettingsStoreMap> {
  if (settingsStoreCache) {
    return settingsStoreCache
  }

  if (!settingsStoreLoadPromise) {
    settingsStoreLoadPromise = readSettingsStoreFile()
      .then((store) => {
        settingsStoreCache = store
        return store
      })
      .finally(() => {
        settingsStoreLoadPromise = null
      })
  }

  return settingsStoreLoadPromise
}

function cloneSettingsStoreMap(store: SettingsStoreMap): SettingsStoreMap {
  return { ...store }
}

async function persistSettingsStore(store: SettingsStoreMap) {
  const filePath = getFrontendStorageFilePath()
  await fs.mkdir(Path.dirname(filePath), { recursive: true })
  await fs.writeFile(filePath, JSON.stringify(store, null, 2), 'utf-8')
}

function enqueueSettingsStoreWrite<T>(task: () => Promise<T>) {
  const run = settingsStoreWriteQueue.then(task, task)
  settingsStoreWriteQueue = run.then(() => {}, () => {})
  return run
}

export async function waitForPendingSettingsStoreWrites() {
  await settingsStoreWriteQueue
}

export async function getSettingsValue(key: string): Promise<SettingsStoreValue | null> {
  await waitForPendingSettingsStoreWrites()
  const store = await ensureSettingsStoreLoaded()
  return store[key] ?? null
}

export async function getAllSettingsValues(): Promise<SettingsStoreMap> {
  await waitForPendingSettingsStoreWrites()
  const store = await ensureSettingsStoreLoaded()
  return cloneSettingsStoreMap(store)
}

export async function reloadSettingsStore(): Promise<ReloadedSettingsStore> {
  await waitForPendingSettingsStoreWrites()
  const previous = settingsStoreCache ? cloneSettingsStoreMap(settingsStoreCache) : cloneSettingsStoreMap(await ensureSettingsStoreLoaded())
  const current = cloneSettingsStoreMap(await readSettingsStoreFile())
  settingsStoreCache = current
  settingsStoreLoadPromise = null
  return {
    previous,
    current,
  }
}

export async function setSettingsValue(key: string, value: SettingsStoreValue): Promise<SettingsStoreValue> {
  return await enqueueSettingsStoreWrite(async () => {
    const currentStore = await ensureSettingsStoreLoaded()
    const nextStore = {
      ...currentStore,
      [key]: value,
    }
    await persistSettingsStore(nextStore)
    settingsStoreCache = nextStore
    return value
  })
}

export async function deleteSettingsValue(key: string): Promise<null> {
  return await enqueueSettingsStoreWrite(async () => {
    const currentStore = await ensureSettingsStoreLoaded()
    if (!(key in currentStore)) {
      return null
    }

    const nextStore = { ...currentStore }
    delete nextStore[key]
    await persistSettingsStore(nextStore)
    settingsStoreCache = nextStore
    return null
  })
}
